### PR TITLE
Implemented warm starts for Lanczos algorithm

### PR DESCRIPTION
**Pull request by shreyashankar**
Run for fewer # iterations by storing the previous eigenvector

### CODE REVIEW

File: cleverhans/experimental/certification/utils.py [hunk 1]
```diff
@@ -183,6 +183,7 @@ def minimum_eigen_vector(x, num_steps, learning_rate, vector_prod_fn):
 
 def tf_lanczos_smallest_eigval(vector_prod_fn,
                                matrix_dim,
+                               b,
```
Comment (AlexeyKurakin) [inline cleverhans/experimental/certification/utils.py]: rename it to something more self-explanatory, like `initial_vector`

File: cleverhans/experimental/certification/utils.py [hunk 2]
```diff
@@ -216,8 +218,11 @@ def tf_lanczos_smallest_eigval(vector_prod_fn,
   q_vectors = tf.TensorArray(
       dtype, size=1, dynamic_size=True, element_shape=(matrix_dim, 1))
 
-  # Create random vector with Euclidean norm 1
-  b = tf.random_normal(shape=(matrix_dim, 1), dtype=dtype)
+  # If start vector is all zeros, make it a random normal vector and run for 1000 iter
+  if tf.reduce_sum(b) == 0:
```
Comment (AlexeyKurakin) [inline cleverhans/experimental/certification/utils.py]: summing up and comparing to zero could be fragile. First of all it's possible that legitimate non-zero b will sum up to 0. Also generally speaking most of floating point computations introduce errors. I'm not sure whether summing up bunch of zeros could be non zero, but better to avoid it just in case.

I would recommend you to do one of the following:

1. compare `b` to None, if it's None then you create new random vector. If it's non None then you use it directly.
   The disadvantage is that you have to call `tf_lanczos_smallest_eigval` twice, one time with b=None and another with b not None, thus you will have two subgraphs.

2. Provide boolean flag to this function and based on value of this flag re-initialize b:
  
```
def tf_lanczos_smallest_eigval(... , use_warm_start, ...):
  
  if use_warm_start == False:   # may need to try few variations of the condition to make autograph happy
    b = tf.random_normal(...)
```

Keep in mind that such `use_warm_start` flag could be Tensor of type `tf.bool`, i.e. `tf.placeholder(tf.bool, shape=())`

3. If you want to actually compare norm of `b` with something rather than having another boolean argument or None, then you should use `tf.norm(b) < TOLERANCE`.

File: cleverhans/experimental/certification/utils.py [hunk 3]
```diff
@@ -216,8 +218,11 @@ def tf_lanczos_smallest_eigval(vector_prod_fn,
...
+    b = tf.random_normal(shape=(matrix_dim, 1), dtype=dtype)
+    max_iter = 1000
```
Comment (AlexeyKurakin) [inline cleverhans/experimental/certification/utils.py]: This value should be an argument of the function.

If you decide to go route 1 (from the comment above) then you don't need this at all.

If you decide to go route 2 (from the comment above) then you can do one of the following
* provide two arguments with number of iterations: `num_iter_full`, `num_iter_warm_start` and choose proper one inside the function
* have only one argument `max_iter`, but when calling this function make it equal to something like `tf.cond(use_warm_starts, lambda: tf.constant(100), lambda: tf.constant(1000))`

If you decide to go route 3, then probably you should have two arguments `num_iter_full`, `num_iter_warm_start`

File: cleverhans/experimental/certification/dual_formulation.py [hunk 4]
```diff
@@ -425,7 +432,13 @@ def compute_certificate(self, current_step):
     lambda_neg_val = self.sess.run(self.lambda_neg)
     lambda_lu_val = self.sess.run(self.lambda_lu)
 
-    _, min_eig_val_h = self.get_lanczos_eig(compute_m=False)
+    # Use 1000 iterations for Lanczos in compute certificate
```
Comment (AlexeyKurakin) [inline cleverhans/experimental/certification/dual_formulation.py]: 1000 should be in self.lzs_params
We may want to adjust specific number

File: cleverhans/experimental/certification/utils.py (hunk 2 above)
Comment (shreyashankar) [reply cleverhans/experimental/certification/utils.py]: I will do #3 because if the norm is anyways less than a tolerance, then we should not be using that vector

File: cleverhans/experimental/certification/dual_formulation.py [hunk 5]
```diff
@@ -90,8 +90,8 @@ def __init__(self, sess, dual_var, neural_net_param_object, test_input, true_cla
       lo_plus_up = self.nn_params.forward_pass(self.lower[i] + self.upper[i], i)
       lo_minus_up = self.nn_params.forward_pass(self.lower[i] - self.upper[i], i, is_abs=True)
       up_minus_lo = self.nn_params.forward_pass(self.upper[i] - self.lower[i], i, is_abs=True)
-      current_lower = 0.5 * (lo_plus_up + lo_minus_up + self.nn_params.biases[i])
-      current_upper = 0.5 * (lo_plus_up + up_minus_lo + self.nn_params.biases[i])
+      current_lower = 0.5 * (lo_plus_up + lo_minus_up) + self.nn_params.biases[i]
+      current_upper = 0.5 * (lo_plus_up + up_minus_lo) + self.nn_params.biases[i]
```
Comment (AlexeyKurakin) [inline cleverhans/experimental/certification/dual_formulation.py]: What is this change about?

Is it the bug in bounds computation which you mentioned before? If yes - it's not really related to this PR and it's better to put it into separate own PR.

File: cleverhans/experimental/certification/utils.py [hunk 6]
```diff
@@ -219,11 +221,11 @@ def tf_lanczos_smallest_eigval(vector_prod_fn,
       dtype, size=1, dynamic_size=True, element_shape=(matrix_dim, 1))
 
   # If start vector is all zeros, make it a random normal vector and run for 1000 iter
-  if tf.reduce_sum(b) == 0:
-    b = tf.random_normal(shape=(matrix_dim, 1), dtype=dtype)
-    max_iter = 1000
+  if tf.norm(initial_vector) < collapse_tol:
```
Comment (AlexeyKurakin) [inline cleverhans/experimental/certification/utils.py]: Please update comment above. It's no longer 1000 iterations, but number specified by an argument

File: cleverhans/experimental/certification/dual_formulation.py (hunk 5 above)
Comment (shreyashankar) [reply cleverhans/experimental/certification/dual_formulation.py]: yes this was the bug. ok
